### PR TITLE
libc: newlib: simplfy malloc arena defn

### PR DESCRIPTION
**Pull request by tyrantlizard77**
We've come a long way since this was written to implement
generic ram bounds definitions and MPU capabilities,
use them here.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

### CODE REVIEW

**Review by dcpleung — APPROVED** [review]

**Comment by andrewboie** [comment]
Oh weird, some network tests like tests/net/socket/udp/net.socket.udp are now build-failing on mps2_an385 due to a check I added:
```
#error "You need to set CONFIG_NEWLIB_LIBC_ALIGNED_HEAP_SIZE"
```
The thing is, I can't see how this worked before. You can't leave it at 0, newlib when it bootstraps itself calls malloc() a few times, having an available malloc() arena is not optional for this libc. Need to dig into this more.

**Review by ioannisg — COMMENTED** [review]
Logic seems fine but as you mention perhaps we need to be setting the CONFIG_NEWLIB_LIBC_ALIGNED_HEAP_SIZE explicitly to a non-zero value

**Comment by andrewboie** [comment]
> 
> 
> Logic seems fine but as you mention perhaps we need to be setting the CONFIG_NEWLIB_LIBC_ALIGNED_HEAP_SIZE explicitly to a non-zero value

I think what's happening here is that the heap arena gets set up from `_end` to the end or RAM, but the bounds of z_malloc_partition get populated with values that won't work if it is inserted from a domain. Since these tests don't add the partition there is no error. The heap still works for supervisor mode access. I'll sort this out, just a bit more cleanup.

File: include/sys/libc-hooks.h [hunk 1]
```diff
@@ -37,10 +37,33 @@ __syscall size_t zephyr_fwrite(const void *_MLIBC_RESTRICT ptr, size_t size,
 #endif /* CONFIG_NEWLIB_LIBC */
 
 #ifdef CONFIG_USERSPACE
-#if defined(CONFIG_NEWLIB_LIBC) || (CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE > 0)
+#if defined(CONFIG_NEWLIB_LIBC)
+/* If we are using newlib, the heap arena is in one of two areas:
+ *  - If we have an MPU that requires power of two alignment, the heap bounds
+ *    must be specified in Kconfig via CONFIG_NEWLIB_LIBC_ALIGNED_HEAP_SIZE.
+ *  - Otherwise, the heap arena is on most arches starts at a suitably
```
Comment (ioannisg) [inline include/sys/libc-hooks.h:44]: remove 'is'

File: include/sys/libc-hooks.h [hunk 2]
```diff
@@ -37,10 +37,33 @@ __syscall size_t zephyr_fwrite(const void *_MLIBC_RESTRICT ptr, size_t size,
...
+ *    aligned base addreess after the `_end` linker symbol, through to the end
```
Comment (ioannisg) [inline include/sys/libc-hooks.h:45]: "to the end", twice written

**Review by ioannisg — CHANGES_REQUESTED** [review]
lgtm, pls, fix the doc issues

File: include/sys/libc-hooks.h (hunk 2 above)
Comment (andrewboie) [reply include/sys/libc-hooks.h:45]: fixed.

File: include/sys/libc-hooks.h (hunk 1 above)
Comment (andrewboie) [reply include/sys/libc-hooks.h]: fixed.

**Review by ioannisg — APPROVED** [review]

File: lib/libc/newlib/libc-hooks.c [hunk 3]
```diff
@@ -30,64 +30,53 @@ MALLOC_BSS static unsigned char __aligned(CONFIG_NEWLIB_LIBC_ALIGNED_HEAP_SIZE)
...
+ * bounds have to go in a k_mem_partition.
  */
-#define HEAP_BASE ((USED_RAM_END_ADDR + \
-		CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE) & \
-	(~(CONFIG_ARM_MPU_REGION_MIN_ALIGN_AND_SIZE - 1)))
+#ifdef CONFIG_MMU
+/* Linker script may already have done this, but just to be safe */
+#define HEAP_BASE	ROUND_UP(USED_RAM_END_ADDR, CONFIG_MMU_PAGE_SIZE)
+#else /* MPU-based systems */
+/* TODO: Need a generic Kconfig for the MPU region granularity */
```
Comment (jenmwms) [inline lib/libc/newlib/libc-hooks.c:49]: This is marked TODO, introduced by this patch. Do we need to create an issue/enhancement for this?

File: lib/libc/newlib/libc-hooks.c [hunk 4]
```diff
@@ -30,64 +30,53 @@ MALLOC_BSS static unsigned char __aligned(CONFIG_NEWLIB_LIBC_ALIGNED_HEAP_SIZE)
...
+#define MAX_HEAP_SIZE  (POINTER_TO_UINT(&_heap_sentry) - HEAP_BASE)
 #else
-#define USED_RAM_SIZE  (USED_RAM_END_ADDR - CONFIG_SRAM_BASE_ADDRESS)
-#define MAX_HEAP_SIZE ((KB(CONFIG_SRAM_SIZE)) - USED_RAM_SIZE)
-#endif
-
-#ifndef HEAP_BASE
-#define HEAP_BASE USED_RAM_END_ADDR
+#define MAX_HEAP_SIZE	(KB(CONFIG_SRAM_SIZE) - \
+			 (HEAP_BASE - CONFIG_SRAM_BASE_ADDRESS))
```
Comment (jenmwms) [inline lib/libc/newlib/libc-hooks.c:69]: Nit: a set of parentheses around the 1st argument was dropped in this consolidation, not sure if it is significant but fwiw - expanding from before this patch would be
`#define MAX_HEAP_SIZE ((KB(CONFIG_SRAM_SIZE)) - (HEAP_BASE - CONFIG_SRAM_BASE_ADDRESS))`

**Review by jenmwms — APPROVED** [review]

File: lib/libc/newlib/libc-hooks.c (hunk 4 above)
Comment (andrewboie) [reply lib/libc/newlib/libc-hooks.c:69]: definition of KB() should already do the right thing wrt to parens, it shouldn't need to itself be enclosed in parens.

File: lib/libc/newlib/libc-hooks.c (hunk 3 above)
Comment (andrewboie) [reply lib/libc/newlib/libc-hooks.c:49]: Yeah, let me file one, good idea.

Comment (andrewboie) [reply lib/libc/newlib/libc-hooks.c:49]: I added this to the scope of https://github.com/zephyrproject-rtos/zephyr/issues/26161

**Pull request merged**